onnx_trt_backend.cpp

```diff
@@ -0,0 +1,771 @@
+  ClearDeviceBuffers();
+  // Setup the input/output bindings and decide batch size
+  for (unsigned i = 0; i < inputsCount; ++i) {
+    if (!inputDescriptors[i].name) {
+      return ONNXIFI_STATUS_INVALID_NAME;
+    }
+    // We only support NCHW
+    if (inputDescriptors[i].dimensions != 4) {
+      return ONNXIFI_STATUS_INVALID_SHAPE;
+    }
+    if (i == 0) {
+      batch_size_ = inputDescriptors[i].shape[0];
+    } else {
+      if (batch_size_ != inputDescriptors[i].shape[0]) {
+        return ONNXIFI_STATUS_INVALID_SHAPE;
+      }
+    }
+    std::cerr << "Adding input " << i << ": " << inputDescriptors[i].name
+              << ", type: " << inputDescriptors[i].memoryType << std::endl;
+    input_map_.emplace(std::string(inputDescriptors[i].name),
+                       inputDescriptors + i);
+  }
+
+  // We don't support the case when batch size is larger than max batch size
+  // yet, but this is not a hard constraint.
+  if (batch_size_ > max_batch_size_) {
+    return ONNXIFI_STATUS_NO_DEVICE_RESOURCES;
+  }
+
+  for (unsigned i = 0; i < outputsCount; ++i) {
+    if (!outputDescriptors[i].name) {
+      return ONNXIFI_STATUS_INVALID_NAME;
+    }
+    // We only support NCHW
+    if (outputDescriptors[i].dimensions != 4) {
+      return ONNXIFI_STATUS_UNSUPPORTED_SHAPE;
+    }
+    if (batch_size_ != outputDescriptors[i].shape[0]) {
+      return ONNXIFI_STATUS_UNSUPPORTED_SHAPE;
+    }
+    output_map_.emplace(std::string(outputDescriptors[i].name),
+                        outputDescriptors + i);
+  }
+
+  int nbindings = trt_engine_->getNbBindings();
+  for (int b = 0; b < nbindings; ++b) {
+    nvinfer1::Dims dims = trt_engine_->getBindingDimensions(b);
+    // Check data type consistency
+    auto binding_datatype = trt_engine_->getBindingDataType(b);
+    if (binding_datatype != nvinfer1::DataType::kFLOAT) {
+      return ONNXIFI_STATUS_MISMATCHING_DATATYPE;
+    }
+
+    if (trt_engine_->bindingIsInput(b)) {
+      std::cerr << "Input: " << trt_engine_->getBindingName(b)
+                << ", Dim: " << dims.d[0] << ", " << dims.d[1] << ", "
+                << dims.d[2] << std::endl;
+      const auto it = input_map_.find(trt_engine_->getBindingName(b));
+      if (it == input_map_.end()) {
+        return ONNXIFI_STATUS_INVALID_NAME;
+      }
+      if (auto ret =
+              CheckAndBindTensor(dims, *it->second) != ONNXIFI_STATUS_SUCCESS) {
+        std::cerr << "Bad binding input" << std::endl;
+        return ret;
+      }
+    } else {
+      // output
+      const auto it = output_map_.find(trt_engine_->getBindingName(b));
+      if (it == output_map_.end()) {
+        return ONNXIFI_STATUS_INVALID_NAME;
+      }
+      if (auto ret =
+              CheckAndBindTensor(dims, *it->second) != ONNXIFI_STATUS_SUCCESS) {
+        return ret;
+      }
+    }
+  }
+
+  trt_executor_ = infer_object(trt_engine_->createExecutionContext());
+  return ONNXIFI_STATUS_SUCCESS;
+}
+
+onnxStatus GraphRep::Run() {
```

Comment:
This isn't really async: you use synchronous copy and call `cudaStreamSynchronize`. To make it really async, you need some changes:

1. Call `cudaEventSynchronize` to make the CUDA stream wait until the input even is signaled.
2. For input tensors in CPU memory, call `cudaMemcpyAsync` to enqueue an operation to the stream to copy them into GPU memory. As this operation is enqueued after `cudaEventSynchronize`, the actual execution wouldn't start until after the input event is signaled.
3. Call `trt_executor_->enqueue(...)`. As this operation is enqueued after `cudaEventSynchronize` and `cudaMemcpyAsync`, the actual execution wouldn't start until after the input event is signaled and all CPU inputs are copied into GPU memory.
4. Do not call `cudaStreamSynchronize`, we want async performance!
5. For output tensors in CPU memory, call `cudaMemcpyAsync` to enqueue an operation to the stream to copy them from GPU memory.
6. Call `cudaEventRecord` to enqueue an operation to signal the output event. The output event will be signaled only after the input event gets signaled, all CPU input are copied to GPU, TensorRT finish inference pass, and all CPU outputs are copied from GPU.

Reply:
`cudaStreamWaitEvent` and `cudaEventRecord` wrapped in event wait()/signal functions. The only things that need to be done is change to `cudaMemcpyAsync` and remove `cudaStreamSynchronize`.  I'll fix it.